cli/cmdlineparser.cpp

```diff
@@ -507,10 +507,17 @@ bool CmdLineParser::parseFromArgs(int argc, const char* const argv[])
                     printError("argument to '-j' is not valid - " +  err + ".");
                     return false;
                 }
-                if (tmp > 10000) {
-                    // This limit is here just to catch typos. If someone has
-                    // need for more jobs, this value should be increased.
-                    printError("argument for '-j' is allowed to be 10000 at max.");
+                if (tmp == 0) {
+                    // TODO: implement get CPU logical core count and use that.
```

Comment:
I do not think we should add such logic in our code. `std::thread::hardware_concurrency()` *might* provide the desired information but it is implementation-dependent and any fallback would be extremely non-portable code which I would like to avoid.

I think it is feasible enough for people to use `nproc`.

Reply:
gmake is open source. Their -j 0 counts logical cores and spawns that many threads, and is about as stable as software gets. Might be worth looking into how they implemented it.

Perhaps it's not as complex as it might seem... I think that's probably C though.

Reply:
I think it would make perfect sense if `-j 0` meant that the logical core count is used (or a sensible fallback, such as `2`).

Using `nproc` isn't portable either, as such a command is not available on e.g. macOS or OpenBSD. Yes, there is another command that works on those two, but that one, in turn, does not work on Linux.

Reply:
> gmake is open source. Their -j 0 counts logical cores and spawns that many threads, and is about as stable as software gets.

No it doesn't. It will schedules *all* possible jobs which would seriously overload your system. That is also reflected in the manpage. You should only ever do that on dedicated (build) systems anyways - but even then it just seems wrong. I think there's actually some serious magic going on with the jobserver with make that does the actual scheduling. There even was a issue in the Linux kernel when you were specifying too many jobs which was encountered and fix by Linus himself after he got his hands on a 100+ core server: https://www.phoronix.com/news/Linux-Pipe-Parallel-Job-Opt.

The jobserver also requires proper Makefiles as it is possible to bypass it leading to unfortunately scheduling. I have never dealt with that myself but I often saw adjustments in the vlc project for that.

Also just copying code from another project is never a good practice. Also there's just more code to maintain (and understand) and we are seriously short on maintainers and some of the existing code is still not quite maintainable.

I think the reasoning for "overscheduling" the system is, that if you match the jobs 1:1 on the actual available threads you  will be slower since not each job will always consume 100% so there's a bit of resources available for those extra jobs.

The scheduling in Cppcheck is currently very bad since the spawning/joining of the process is sequential. And there's also some overhead in preparing in job and transferring the data to the main process. That causes serious overhead/stalling when you have a lot of small files to scan. I have been working on this for quite a while now and have a hacked version which behaves as it should but getting those changes in is incremental and sometimes not very straight forward. I also have not been able to work that much on it. But getting closer. 🤞

> Using nproc isn't portable either, as such a command is not available on e.g. macOS or OpenBSD. 

`nproc` is available on macOS via `brew coreutils` (it's a must anyways since common commands just don't work as expected on MacOS).

And a very weak closing argument: In the 10+ years of this project (not exactly sure when this option was added) nobody seems to have encountered this so it doesn't seem important/necessary. 😉

Reply:
> No it doesn't. It will schedules _all_ possible jobs which would seriously overload your system. That is also reflected in the manpage.

My mistake. It does indeed say it "will not limit the number of jobs." I must have been thinking of another tool. Not sure I would say it overloads your system though; I used -j 0 in build scripts all the time on macOS and Linux and never had a problem, unless the project has a dependency / concurrency issue, which I've seen. In that case, you *have* to use a single thread.

I don't presume to know anything about cppcheck's job creation process; I only glanced through while submitting this PR, so I can't comment on that.

I can see the argument for creating more threads than logical cores available if those jobs are I/O bound instead of CPU bound. Makes sense. It would be fun to generate performance numbers with different size and type of project (huge files, tiny files, a ton of files, just a few, etc.) and graph it out and see what the best average would look like.

I work full-time and do my own things on the side, but I like cppcheck, so maybe I could find time to help here and there if you guys need more people...